ftw/testbrowser/drivers/traversaldriver.py

```diff
@@ -0,0 +1,330 @@
+from App.ZApplication import ZApplicationWrapper
+from ftw.testbrowser.drivers.utils import isolated
+from ftw.testbrowser.drivers.utils import remembering_for_reload
+from ftw.testbrowser.exceptions import BlankPage
+from ftw.testbrowser.exceptions import RedirectLoopException
+from ftw.testbrowser.interfaces import IDriver
+from ftw.testbrowser.utils import copy_docs_from_interface
+from requests.structures import CaseInsensitiveDict
+from StringIO import StringIO
+from urllib import unquote
+from urlparse import urlparse
+from zope.interface import implements
+from ZPublisher.BaseRequest import RequestContainer
+from ZPublisher.Iterators import IStreamIterator
+from ZPublisher.Response import Response
+from ZPublisher.Test import publish_module
+import httplib
+import requests
+import sys
+import Zope2
+import ZPublisher
+
+
+# from plone.testing._z2_testbrowser
+class TestResponse(Response):
+
+    def setBody(self, body, title='', is_error=0, **kw):
+        if IStreamIterator.providedBy(body):
+            body = ''.join(body)
+        Response.setBody(self, body, title, is_error, **kw)
+
+
+class NoCommitTransactionsManagerWrapper(object):
+    """Wraps a Zope2.App.startup.TransactionsManager and patches
+    Zope2.zpublisher_transactions_manager with a singleton instance.
+
+    The wrapper lives for the rest of the process duration since it
+    is cached in the function argument default dict ``modules`` of the
+    function ``ZPublisher.Publish.get_module_info``.
+
+    When the wrapper is activated with ``activate_wrapper`` it silently
+    aborts all commit intents.
+    When then transaction is beginning, we replace the ZApplicationWrapper
+    object in the requests's PARENTS with an actual application root with
+    the current connection in order to avoid a new connection to be created
+    when the ZApplicationWrapper is traversed, since this would result in a
+    reset of uncommitted state.
+    """
+
+    @classmethod
+    def setup(klass):
+        """Make sure the wrapper is active and return the wrapper.
+        May or may not create a new wrapper.
+        """
+        if not isinstance(Zope2.zpublisher_transactions_manager, klass):
+            Zope2.zpublisher_transactions_manager = klass(
+                Zope2.zpublisher_transactions_manager)
+
+            modules_cache = ZPublisher.Publish.get_module_info.func_defaults[0]
+            assert isinstance(modules_cache, dict), \
+                'get_module_info modules cache changed unexpectedly.'
+            modules_cache.pop('Zope2', None)
+            modules_cache.pop('Zope2.cgi', None)
+
+        return Zope2.zpublisher_transactions_manager
+
+    def __init__(self, wrapped):
+        self.wrapped = wrapped
+        self.active = False
+        self.next_request = None
+        self.current_app = None
+
+    def activate_wrapper(self, next_request, current_app):
+        self.active = True
+        self.next_request = next_request
```

Comment:
i don't really get what `next_request` and `current_app` are here for 😢 , could you add a docstring?